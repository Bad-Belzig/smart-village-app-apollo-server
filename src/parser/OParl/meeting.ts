import { isObjectLike } from 'lodash';

import { parseDate } from '../dateParser';
import { mapToIds } from '../parserHelpers';

// only the necessary validation for parsing is done here.
// the created database object will be validated after creation before saving
// by using the .validate method that checks against the provided schema
export const parseMeeting = (json) => {
  if (isObjectLike(json)) {
    const externalId = json.id; // change the id given to us to be the externalId instead
    delete json.id; // remove the id key from the json to ensure we do not preset it for mongo db

    return {
      ...json,
      externalId,
      created: parseDate(json.created),
      modified: parseDate(json.modified),

      start: parseDate(json.start),
      end: parseDate(json.end),

      agendaItem: mapToIds(json.agendaItem),
      auxiliaryFile: mapToIds(json.auxiliaryFile),
      invitation: json.invitation?.id,
      location: json.location?.id,
      resultsProtocol: json.resultsProtocol?.id,
      verbatimProtocol: json.verbatimProtocol?.id,
    };
  }
};
