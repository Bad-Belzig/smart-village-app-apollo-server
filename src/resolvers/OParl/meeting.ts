import {
  AgendaItem,
  File,
  IMeeting,
  Location,
  Meeting,
  Organization,
  Person,
} from '../../models';
import { findByIds, getPaginatedEntriesByIds } from '../resolverHelpers';

export const meetingResolvers = {
  Query: {
    oParlMeetings: (
      _,
      args: { externalIds?: string[]; before?: string; after?: string },
    ) => {
      const dateFilter: { $gte?: Date; $lt?: Date } = {};
      if (args.after) dateFilter.$gte = new Date(args.after);
      if (args.before) dateFilter.$lt = new Date(args.before);

      return args.externalIds
        ? findByIds(args.externalIds, Meeting)
        : Meeting.find({
            start: dateFilter,
          });
    },
  },
  OParlMeeting: {
    location: (parent: IMeeting) =>
      Location.findOne({ externalId: parent.location }),
    organization: async (
      parent: IMeeting,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Organization,
        parent.organization,
        args.offset,
        args.pageSize,
      ),
    participant: async (
      parent: IMeeting,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        Person,
        parent.participant,
        args.offset,
        args.pageSize,
      ),
    invitation: (parent: IMeeting) =>
      File.findOne({ externalId: parent.invitation }),
    resultsProtocol: (parent: IMeeting) =>
      File.findOne({ externalId: parent.resultsProtocol }),
    verbatimProtocol: (parent: IMeeting) =>
      File.findOne({ externalId: parent.verbatimProtocol }),
    auxiliaryFile: async (
      parent: IMeeting,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        File,
        parent.auxiliaryFile,
        args.offset,
        args.pageSize,
      ),
    agendaItem: async (
      parent: IMeeting,
      args: { offset?: number; pageSize?: number },
    ) =>
      getPaginatedEntriesByIds(
        AgendaItem,
        parent.agendaItem,
        args.offset,
        args.pageSize,
      ),
  },
};
