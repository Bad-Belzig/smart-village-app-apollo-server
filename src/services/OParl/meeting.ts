import { fetch } from 'apollo-env';

import { Meeting } from '../../models';
import { parseMeeting } from '../../parser';

// import flow
// 1. fetch json
// 2. create db object
//   2.1. validation necessary for parsing (done in parser)
//   2.2. parse json
//   2.3. create object
// 3. validate db object
// 4. queue new related objects
// 5. save to db

export const importMeeting = async (url: string) => {
  try {
    const response = await fetch(url);

    if (response.ok) {
      const json = await response.json();

      // TODO: check for previous existence

      const lt = new Meeting(parseMeeting(json));

      await lt.validate();

      // TODO: check for persons, organizations, files and location to import

      return lt.save();
    } else {
      throw new Error(
        `Error while fetching Meeting from : ${url}! \n response.status: ${response.status}`,
      );
    }
  } catch (e) {
    console.log(e);
  }
};
