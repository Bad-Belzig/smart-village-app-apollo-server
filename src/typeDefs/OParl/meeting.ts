import { gql } from 'apollo-server-core';

export const meetingGQL = gql`
  extend type Query {
    oParlMeetings: [OParlMeeting!]
  }

  type OParlMeeting {
    id: ID!
    "id from the oparl object = url where the json is located"
    externalId: String!
    "oParl type = url where the type is specified"
    type: String!
    created: Date
    modified: Date
    keyword: [String!]
    web: String
    deleted: Boolean
    agendaItem: [OParlAgendaItem!]
    auxiliaryFile: [OParlFile!]
    cancelled: Boolean
    end: Date
    invitation: OParlFile
    location: OParlLocation
    meetingState: String
    name: String
    organization: [OParlOrganization!]
    participant: [OParlPerson!]
    resultsProtocol: OParlFile
    start: Date
    verbatimProtocol: OParlFile
  }
`;
