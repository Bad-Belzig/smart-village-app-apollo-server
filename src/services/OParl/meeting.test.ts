import {
  basicImportTest,
  invalidUrlTest,
  setupTestDataBase,
} from '../../helpers';
import { importMeeting } from './meeting';

const helperFuncs = setupTestDataBase('test_meeting_import');

// TODO: change test url to something within our control
const testUrl =
  'https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/meeting/503';

describe('importing a meeting', () => {
  beforeAll(helperFuncs.beforeAll);

  it('imports properly', basicImportTest(importMeeting, testUrl));

  it('import returns undefined on faulty url', invalidUrlTest(importMeeting));

  afterAll(helperFuncs.afterAll);
});
