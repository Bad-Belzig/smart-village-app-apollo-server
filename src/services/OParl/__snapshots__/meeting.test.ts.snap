// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importing a meeting imports properly 1`] = `
Object {
  "__v": 0,
  "agendaItem": Array [
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/4155",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/317",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/325",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/326",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/327",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/328",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/329",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/330",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/331",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/320",
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/agendaitem/299",
  ],
  "auxiliaryFile": Array [],
  "created": 2017-07-24T15:33:12.000Z,
  "end": 2004-11-16T19:00:00.000Z,
  "externalId": "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/meeting/503",
  "keyword": Array [
    "Rat der Gemeinde",
  ],
  "modified": 2017-08-01T12:14:43.000Z,
  "name": "Rat der Gemeinde",
  "organization": Array [
    "https://ris.schwalmtal.de/webservice/oparl/v1.0/body/1/organization/1-1",
  ],
  "participant": Array [],
  "start": 2004-11-16T18:00:00.000Z,
  "type": "https://schema.oparl.org/1.0/Meeting",
  "web": "https://ris.schwalmtal.de/tops/?__=UGhVM0hpd2NXNFdFcExjZcG8mlSHys8pcEOe79OBc6U",
}
`;
